[FEATURE] Add support for number of documents when creating workloads

**Is your feature request related to a problem? Please describe.**
`create-workload` subcommand currently generates a workload using all the documents from appointed indices. It would be nice if users could specify a number of documents from the indices they've appointed instead of using all the documents.

Currently, this is what `create-workload` supports:
```
usage: opensearch-benchmark create-workload [-h] --workload WORKLOAD --indices INDICES --target-hosts TARGET_HOSTS [--client-options CLIENT_OPTIONS]
                                            [--output-path OUTPUT_PATH] [--quiet] [--offline]

options:
  -h, --help            show this help message and exit
  --workload WORKLOAD   Name of the generated workload
  --indices INDICES     Comma-separated list of indices to include in the workload
  --target-hosts TARGET_HOSTS
                        Comma-separated list of host:port pairs which should be targeted
  --client-options CLIENT_OPTIONS
                        Comma-separated list of client options to use. (default: timeout:60)
  --output-path OUTPUT_PATH
                        Workload output directory (default: workloads/)
  --quiet               Suppress as much as output as possible (default: false).
  --offline             Assume that Benchmark has no connection to the Internet (default: false).
```

**Describe the solution you'd like**
There should be an additional parameter that takes in the number of documents of appointed indices to generate workload from.

**Describe alternatives you've considered**

**Additional context**

Add any other context or screenshots about the feature request here.

### Using a single index and single doc count
```
hoangia@3c22fbd0d988 opensearch-benchmark % opensearch-benchmark create-workload --workload=movies-test-with-parameter --target-hosts=<TARGET HOST ENDPOINT> --client-options="basic_auth_user:'<USERNAME>',basic_auth_password:'<PASSWORD>'" --indices=movies-2000 --output-path=~/Desktop/workloads --total-docs=1200

   ____                  _____                      __       ____                  __                         __
  / __ \____  ___  ____ / ___/___  ____ ___________/ /_     / __ )___  ____  _____/ /_  ____ ___  ____ ______/ /__
 / / / / __ \/ _ \/ __ \\__ \/ _ \/ __ `/ ___/ ___/ __ \   / __  / _ \/ __ \/ ___/ __ \/ __ `__ \/ __ `/ ___/ //_/
/ /_/ / /_/ /  __/ / / /__/ /  __/ /_/ / /  / /__/ / / /  / /_/ /  __/ / / / /__/ / / / / / / / / /_/ / /  / ,<
\____/ .___/\___/_/ /_/____/\___/\__,_/_/   \___/_/ /_/  /_____/\___/_/ /_/\___/_/ /_/_/ /_/ /_/\__,_/_/  /_/|_|
    /_/

[INFO] You did not provide an explicit timeout in the client options. Assuming default of 10 seconds.
[INFO] Connected to OpenSearch cluster [eef3f5a9032ea6572cf946f1fbcec586] version [1.1.0].

Extracting documents for index [movies-2000] for test mode... 1000/1000 docs [100.0% done]
Extracting documents for index [movies-2000]...               1200/1200 docs [100.0% done]

[INFO] Workload movies-test-with-parameter has been created. Run it with: opensearch-benchmark --workload-path=/Users/hoangia/Desktop/workloads/movies-test-with-parameter

-------------------------------
[INFO] SUCCESS (took 3 seconds)
-------------------------------
```

### Use multiple indices and multiple doc counts
```
hoangia@3c22fbd0d988 opensearch-benchmark % opensearch-benchmark create-workload --workload=movies-test-with-parameter --target-hos<TARGET HOST ENDPOINT> --client-options="basic_auth_user:'<USERNAME>',basic_auth_password:'<PASSWORD>'" --indices=movies-2000,movies-1000 --output-path=~/Desktop/workloads --total-docs=1200,1000

   ____                  _____                      __       ____                  __                         __
  / __ \____  ___  ____ / ___/___  ____ ___________/ /_     / __ )___  ____  _____/ /_  ____ ___  ____ ______/ /__
 / / / / __ \/ _ \/ __ \\__ \/ _ \/ __ `/ ___/ ___/ __ \   / __  / _ \/ __ \/ ___/ __ \/ __ `__ \/ __ `/ ___/ //_/
/ /_/ / /_/ /  __/ / / /__/ /  __/ /_/ / /  / /__/ / / /  / /_/ /  __/ / / / /__/ / / / / / / / / /_/ / /  / ,<
\____/ .___/\___/_/ /_/____/\___/\__,_/_/   \___/_/ /_/  /_____/\___/_/ /_/\___/_/ /_/_/ /_/ /_/\__,_/_/  /_/|_|
    /_/

[INFO] You did not provide an explicit timeout in the client options. Assuming default of 10 seconds.
[INFO] Connected to OpenSearch cluster [48bd5a55847accbfaebf1c57cfa2e513] version [1.1.0].

Extracting documents for index [movies-2000] for test mode... 1000/1000 docs [100.0% done]
Extracting documents for index [movies-2000]...               1200/1200 docs [100.0% done]
Extracting documents for index [movies-1000] for test mode... 1000/1000 docs [100.0% done]
Extracting documents for index [movies-1000]...               1000/1000 docs [100.0% done]

[INFO] Workload movies-test-with-parameter has been created. Run it with: opensearch-benchmark --workload-path=/Users/hoangia/Desktop/workloads/movies-test-with-parameter

-------------------------------
[INFO] SUCCESS (took 3 seconds)
-------------------------------
```

### When user specifies `--total-docs` but only provides a fraction of indices total docs count
```
hoangia@3c22fbd0d988 opensearch-benchmark % opensearch-benchmark create-workload --workload=movies-test-with-parameter --target-hosts=<TARGET HOST ENDPOINT> --client-options="basic_auth_user:'<USERNAME>',basic_auth_password:'<PASSWORD>'" --indices=movies-2000,movies-1000 --output-path=~/Desktop/workloads --total-docs=1200

   ____                  _____                      __       ____                  __                         __
  / __ \____  ___  ____ / ___/___  ____ ___________/ /_     / __ )___  ____  _____/ /_  ____ ___  ____ ______/ /__
 / / / / __ \/ _ \/ __ \\__ \/ _ \/ __ `/ ___/ ___/ __ \   / __  / _ \/ __ \/ ___/ __ \/ __ `__ \/ __ `/ ___/ //_/
/ /_/ / /_/ /  __/ / / /__/ /  __/ /_/ / /  / /__/ / / /  / /_/ /  __/ / / / /__/ / / / / / / / / /_/ / /  / ,<
\____/ .___/\___/_/ /_/____/\___/\__,_/_/   \___/_/ /_/  /_____/\___/_/ /_/\___/_/ /_/_/ /_/ /_/\__,_/_/  /_/|_|
    /_/

[INFO] You did not provide an explicit timeout in the client options. Assuming default of 10 seconds.
[INFO] Connected to OpenSearch cluster [380d8fd64dd85b5f77c0ad81b0799e1e] version [1.1.0].

Extracting documents for index [movies-2000] for test mode... 1000/1000 docs [100.0% done]
Extracting documents for index [movies-2000]...               1200/1200 docs [100.0% done]
[ERROR] Cannot create-workload. --total-docs missing document counts for corresponding indices.

Getting further help:
*********************
* Check the log files in /Users/hoangia/.benchmark/logs for errors.
* Read the documentation at https://opensearch.org/docs.
* Ask a question on the forum at https://forum.opensearch.org/.
* Raise an issue at https://github.com/opensearch-project/OpenSearch-Benchmark/issues and include the log files in /Users/hoangia/.benchmark/logs.

-------------------------------
[INFO] FAILURE (took 2 seconds)
-------------------------------
```
### If user does not specify doc counts, then it uses all indices in the index. 
```
hoangia@3c22fbd0d988 opensearch-benchmark % opensearch-benchmark create-workload --workload=movies-test-with-parameter --target-hosts=https://search-hoangia-aos-testing-2-xpwa7ykgv7bbc7qqfuuzt2no7i.us-east-1.es.amazonaws.com --client-options="basic_auth_user:'hoangia',basic_auth_password:'Hoangia@123'" --indices=movies-2000,movies-1000 --output-path=~/Desktop/workloads

   ____                  _____                      __       ____                  __                         __
  / __ \____  ___  ____ / ___/___  ____ ___________/ /_     / __ )___  ____  _____/ /_  ____ ___  ____ ______/ /__
 / / / / __ \/ _ \/ __ \\__ \/ _ \/ __ `/ ___/ ___/ __ \   / __  / _ \/ __ \/ ___/ __ \/ __ `__ \/ __ `/ ___/ //_/
/ /_/ / /_/ /  __/ / / /__/ /  __/ /_/ / /  / /__/ / / /  / /_/ /  __/ / / / /__/ / / / / / / / / /_/ / /  / ,<
\____/ .___/\___/_/ /_/____/\___/\__,_/_/   \___/_/ /_/  /_____/\___/_/ /_/\___/_/ /_/_/ /_/ /_/\__,_/_/  /_/|_|
    /_/

[INFO] You did not provide an explicit timeout in the client options. Assuming default of 10 seconds.
[INFO] Connected to OpenSearch cluster [380d8fd64dd85b5f77c0ad81b0799e1e] version [1.1.0].

Extracting documents for index [movies-2000] for test mode... 1000/1000 docs [100.0% done]
Extracting documents for index [movies-2000]...               2000/2000 docs [100.0% done]
Extracting documents for index [movies-1000] for test mode... 1000/1000 docs [100.0% done]
Extracting documents for index [movies-1000]...               1000/1000 docs [100.0% done]

[INFO] Workload movies-test-with-parameter has been created. Run it with: opensearch-benchmark --workload-path=/Users/hoangia/Desktop/workloads/movies-test-with-parameter

-------------------------------
[INFO] SUCCESS (took 3 seconds)
-------------------------------
```

--- a/osbenchmark/benchmark.py
+++ b/osbenchmark/benchmark.py
@@ -182,8 +182,10 @@ def create_arg_parser():
         help="Input JSON file to use containing custom workload queries that override the default match_all query")
     create_workload_parser.add_argument(
         "--number-of-docs",
-        type=opts.csv_to_list,
-        help="Comma-separated list of doc counts for each index specified in indices parameter. Ensure that order is preserved.")
+        action=opts.StoreKeyPairAsDict,
+        nargs='+',
+        metavar="KEY:VAL",
+        help="Map of index name and doc count to extract. Ensure that index name in key field exists in --indices parameter.")
 
     generate_parser = subparsers.add_parser("generate", help="Generate artifacts")
     generate_parser.add_argument(

--- a/osbenchmark/benchmark.py
+++ b/osbenchmark/benchmark.py
@@ -55,12 +55,6 @@ def create_arg_parser():
             raise argparse.ArgumentError(argument=None, message="At least one argument required!")
         return lst
 
-    def list_or_none(arg):
-        lst = opts.csv_to_list(arg)
-        if len(lst) < 1:
-            return None
-        return lst
-
     def runtime_jdk(v):
         if v == "bundled":
             return v
@@ -185,11 +179,10 @@ def create_arg_parser():
     create_workload_parser.add_argument(
         "--custom-queries",
         type=argparse.FileType('r'),
-        help="Input JSON file to use containing custom workload queries that override the default match_all query"
-    )
+        help="Input JSON file to use containing custom workload queries that override the default match_all query")
     create_workload_parser.add_argument(
-        "--total-docs",
-        type=list_or_none,
+        "--number-of-docs",
+        type=opts.csv_to_list,
         help="Comma-separated list of doc counts for each index specified in indices parameter. Ensure that order is preserved.")
 
     generate_parser = subparsers.add_parser("generate", help="Generate artifacts")
@@ -917,7 +910,7 @@ def dispatch_sub_command(arg_parser, args, cfg):
             generate(cfg)
         elif sub_command == "create-workload":
             cfg.add(config.Scope.applicationOverride, "generator", "indices", args.indices)
-            cfg.add(config.Scope.applicationOverride, "generator", "total_docs", args.total_docs)
+            cfg.add(config.Scope.applicationOverride, "generator", "number_of_docs", args.number_of_docs)
             cfg.add(config.Scope.applicationOverride, "generator", "output.path", args.output_path)
             cfg.add(config.Scope.applicationOverride, "workload", "workload.name", args.workload)
             cfg.add(config.Scope.applicationOverride, "workload", "custom_queries", args.custom_queries)

--- a/osbenchmark/benchmark.py
+++ b/osbenchmark/benchmark.py
@@ -55,6 +55,12 @@ def create_arg_parser():
             raise argparse.ArgumentError(argument=None, message="At least one argument required!")
         return lst
 
+    def list_or_none(arg):
+        lst = opts.csv_to_list(arg)
+        if len(lst) < 1:
+            return None
+        return lst
+
     def runtime_jdk(v):
         if v == "bundled":
             return v
@@ -181,6 +187,10 @@ def create_arg_parser():
         type=argparse.FileType('r'),
         help="Input JSON file to use containing custom workload queries that override the default match_all query"
     )
+    create_workload_parser.add_argument(
+        "--total-docs",
+        type=list_or_none,
+        help="Comma-separated list of doc counts for each index specified in indices parameter. Ensure that order is preserved.")
 
     generate_parser = subparsers.add_parser("generate", help="Generate artifacts")
     generate_parser.add_argument(
@@ -907,6 +917,7 @@ def dispatch_sub_command(arg_parser, args, cfg):
             generate(cfg)
         elif sub_command == "create-workload":
             cfg.add(config.Scope.applicationOverride, "generator", "indices", args.indices)
+            cfg.add(config.Scope.applicationOverride, "generator", "total_docs", args.total_docs)
             cfg.add(config.Scope.applicationOverride, "generator", "output.path", args.output_path)
             cfg.add(config.Scope.applicationOverride, "workload", "workload.name", args.workload)
             cfg.add(config.Scope.applicationOverride, "workload", "custom_queries", args.custom_queries)

--- a/osbenchmark/resources/custom-query-workload.json.j2
+++ b/osbenchmark/resources/custom-query-workload.json.j2
@@ -6,6 +6,7 @@
       "operation": {
         "name": "{{query.name}}",
         "operation-type": "{{query['operation-type']}}",
+        "index": {{ indices | map(attribute='name') | list | join(',') | tojson }},
         "body": {{query.body | replace("'", '"') }}
       }
     }{% if not loop.last %},{% endif -%}

--- a/osbenchmark/resources/default-query-workload.json.j2
+++ b/osbenchmark/resources/default-query-workload.json.j2
@@ -4,6 +4,7 @@
     {
       "operation": {
         "operation-type": "search",
+        "index": {{ indices | map(attribute='name') | list | join(',') | tojson }},
         "body": {
           "query": {
             "match_all": {}

--- a/osbenchmark/utils/opts.py
+++ b/osbenchmark/utils/opts.py
@@ -24,6 +24,7 @@
 
 import difflib
 import json
+import argparse
 
 from osbenchmark.utils import io
 
@@ -35,7 +36,6 @@ def csv_to_list(csv):
         return []
     else:
         return [e.strip() for e in csv.split(",")]
-
 
 def to_bool(v):
     if v is None:
@@ -116,6 +116,24 @@ def make_list_of_close_matches(word_list, all_possibilities):
             close_matches.append(matched_word[0])
 
     return close_matches
+
+class StoreKeyPairAsDict(argparse.Action):
+     """
+     Custom Argparse action that allows users to pass in a key:value pairs after specifying a parameter.
+     Used as action for --number-of-docs parameter for create-workload subcommand.
+     """
+     def __init__(self, option_strings, dest, nargs=None, **kwargs):
+         self._nargs = nargs
+         super(StoreKeyPairAsDict, self).__init__(option_strings, dest, nargs=nargs, **kwargs)
+
+     def __call__(self, parser, namespace, values, option_string=None):
+         my_dict = {}
+         for kv in values:
+             k,v = kv.split(":")
+             my_dict[k] = v
+         setattr(namespace, self.dest, my_dict)
+
+         return my_dict
 
 
 class ConnectOptions:

--- a/osbenchmark/utils/opts.py
+++ b/osbenchmark/utils/opts.py
@@ -118,22 +118,22 @@ def make_list_of_close_matches(word_list, all_possibilities):
     return close_matches
 
 class StoreKeyPairAsDict(argparse.Action):
-     """
-     Custom Argparse action that allows users to pass in a key:value pairs after specifying a parameter.
-     Used as action for --number-of-docs parameter for create-workload subcommand.
-     """
-     def __init__(self, option_strings, dest, nargs=None, **kwargs):
-         self._nargs = nargs
-         super(StoreKeyPairAsDict, self).__init__(option_strings, dest, nargs=nargs, **kwargs)
+    """
+    Custom Argparse action that allows users to pass in a key:value pairs after specifying a parameter.
+    Used as action for --number-of-docs parameter for create-workload subcommand.
+    """
+    def __init__(self, option_strings, dest, nargs=None, **kwargs):
+        self._nargs = nargs
+        super().__init__(option_strings, dest, nargs=nargs, **kwargs)
 
-     def __call__(self, parser, namespace, values, option_string=None):
-         my_dict = {}
-         for kv in values:
-             k,v = kv.split(":")
-             my_dict[k] = v
-         setattr(namespace, self.dest, my_dict)
+    def __call__(self, parser, namespace, values, option_string=None):
+        my_dict = {}
+        for kv in values:
+            k,v = kv.split(":")
+            my_dict[k] = v
+        setattr(namespace, self.dest, my_dict)
 
-         return my_dict
+        return my_dict
 
 
 class ConnectOptions:

--- a/osbenchmark/workload_generator/corpus.py
+++ b/osbenchmark/workload_generator/corpus.py
@@ -50,7 +50,7 @@ def get_doc_outpath(outdir, name, suffix=""):
     return os.path.join(outdir, f"{name}-documents{suffix}.json")
 
 
-def extract(client, output_path, index):
+def extract(client, output_path, index, total_docs_requested=None):
     """
     Scroll an index with a match-all query, dumping document source to ``outdir/documents.json``.
 
@@ -62,7 +62,14 @@ def extract(client, output_path, index):
 
     logger = logging.getLogger(__name__)
 
-    total_docs = client.count(index=index)["count"]
+    total_docs_found = client.count(index=index)["count"]
+    if total_docs_requested is None:
+        # If none, just use all documents found
+        total_docs_requested = total_docs_found
+
+    # If total_docs_requested is >= than total_docs_found, it will just use total_docs_found. Other wise, use total_docs_requested
+    total_docs = min(total_docs_requested, total_docs_found)
+
     if total_docs > 0:
         logger.info("[%d] total docs in index [%s].", total_docs, index)
         docs_path = get_doc_outpath(output_path, index)

--- a/osbenchmark/workload_generator/corpus.py
+++ b/osbenchmark/workload_generator/corpus.py
@@ -50,7 +50,7 @@ def get_doc_outpath(outdir, name, suffix=""):
     return os.path.join(outdir, f"{name}-documents{suffix}.json")
 
 
-def extract(client, output_path, index, total_docs_requested=None):
+def extract(client, output_path, index, number_of_docs_requested=None):
     """
     Scroll an index with a match-all query, dumping document source to ``outdir/documents.json``.
 
@@ -62,31 +62,27 @@ def extract(client, output_path, index, total_docs_requested=None):
 
     logger = logging.getLogger(__name__)
 
-    total_docs_found = client.count(index=index)["count"]
-    if total_docs_requested is None:
-        # If none, just use all documents found
-        total_docs_requested = total_docs_found
+    number_of_docs = client.count(index=index)["count"]
 
-    # If total_docs_requested is >= than total_docs_found, it will just use total_docs_found. Other wise, use total_docs_requested
-    total_docs = min(total_docs_requested, total_docs_found)
+    ndocs_to_extract = number_of_docs if not number_of_docs_requested else min(number_of_docs, number_of_docs_requested)
 
-    if total_docs > 0:
-        logger.info("[%d] total docs in index [%s].", total_docs, index)
+    if ndocs_to_extract > 0:
+        logger.info("[%d] total docs in index [%s].", number_of_docs, index)
         docs_path = get_doc_outpath(output_path, index)
-        dump_documents(client, index, get_doc_outpath(output_path, index, "-1k"), min(total_docs, 1000), " for test mode")
-        dump_documents(client, index, docs_path, total_docs)
-        return template_vars(index, docs_path, total_docs)
+        dump_documents(client, index, get_doc_outpath(output_path, index, "-1k"), min(ndocs_to_extract, 1000), " for test mode")
+        dump_documents(client, index, docs_path, ndocs_to_extract)
+        return template_vars(index, docs_path, ndocs_to_extract)
     else:
         logger.info("Skipping corpus extraction fo index [%s] as it contains no documents.", index)
         return None
 
 
-def dump_documents(client, index, out_path, total_docs, progress_message_suffix=""):
+def dump_documents(client, index, out_path, number_of_docs, progress_message_suffix=""):
     # pylint: disable=import-outside-toplevel
     from opensearchpy import helpers
 
     logger = logging.getLogger(__name__)
-    freq = max(1, total_docs // 1000)
+    freq = max(1, number_of_docs // 1000)
 
     progress = console.progress()
     compressor = DOCS_COMPRESSOR()
@@ -96,14 +92,14 @@ def dump_documents(client, index, out_path, total_docs, progress_message_suffix=
             logger.info("Dumping corpus for index [%s] to [%s].", index, out_path)
             query = {"query": {"match_all": {}}}
             for n, doc in enumerate(helpers.scan(client, query=query, index=index)):
-                if n >= total_docs:
+                if n >= number_of_docs:
                     break
                 data = (json.dumps(doc["_source"], separators=(",", ":")) + "\n").encode("utf-8")
 
                 outfile.write(data)
                 comp_outfile.write(compressor.compress(data))
 
-                render_progress(progress, progress_message_suffix, index, n + 1, total_docs, freq)
+                render_progress(progress, progress_message_suffix, index, n + 1, number_of_docs, freq)
 
             comp_outfile.write(compressor.flush())
     progress.finish()

--- a/osbenchmark/workload_generator/workload_generator.py
+++ b/osbenchmark/workload_generator/workload_generator.py
@@ -44,14 +44,12 @@ def process_template(templates_path, template_filename, template_vars, output_pa
 
 def check_if_indices_and_number_of_docs_match(indices, number_of_docs, docs_were_requested):
     '''
-    if --number-of-docs were provided, the input length needs to be match indices.
-
-    Valid examples: --indices=movies,actors --number-of-docs=1000,2000
-    Invalid examples: --indices=movies,actors --number-of-docs=1000 OR
-    --indices=movies --number-of-docs=1000,2000
+    if key(s) and value(s) were provided to --number-of-docs,
+    the number of pairs need to be less than or equal to number of indices
     '''
-    if docs_were_requested and len(indices) != len(number_of_docs):
-        raise exceptions.SystemSetupError("Both --indices and --number-of-docs were provided but missing indices or docs.")
+    if docs_were_requested and len(indices) < len(number_of_docs):
+        raise exceptions.SystemSetupError(f"Number of key:value pairs exceeds number of indices mentioned in --indices. "
+                                          f"Ensure it is less than or equal to.")
 
 def extract_mappings_and_corpora(client, output_path, indices_to_extract, number_of_docs_requested):
     indices = []
@@ -69,8 +67,10 @@ def extract_mappings_and_corpora(client, output_path, indices_to_extract, number
             logging.getLogger(__name__).exception("Failed to extract index [%s]", index_name)
 
     # That list only contains valid indices (with index patterns already resolved)
-    for count, idx in enumerate(indices):
-        c = corpus.extract(client, output_path, idx["name"], int(number_of_docs_requested[count]) if docs_were_requested else None)
+    for i in indices:
+        docs_to_extract = number_of_docs_requested.get(i["name"], None) if docs_were_requested else None
+        docs_to_extract = int(docs_to_extract) if docs_to_extract is not None else docs_to_extract
+        c = corpus.extract(client, output_path, i["name"], docs_to_extract)
         if c:
             corpora.append(c)
 

--- a/osbenchmark/workload_generator/workload_generator.py
+++ b/osbenchmark/workload_generator/workload_generator.py
@@ -51,7 +51,7 @@ def check_if_indices_and_number_of_docs_match(indices, number_of_docs, docs_were
     --indices=movies --number-of-docs=1000,2000
     '''
     if docs_were_requested and len(indices) != len(number_of_docs):
-        raise exceptions.SystemSetupError("Both --indices and --number-of-docs were provided but missing corresponding indices or total docs.")
+        raise exceptions.SystemSetupError("Both --indices and --number-of-docs were provided but missing indices or docs.")
 
 def extract_mappings_and_corpora(client, output_path, indices_to_extract, number_of_docs_requested):
     indices = []
@@ -98,10 +98,10 @@ def create_workload(cfg):
     root_path = cfg.opts("generator", "output.path")
     target_hosts = cfg.opts("client", "hosts")
     client_options = cfg.opts("client", "options")
+    number_of_docs = cfg.opts("generator", "number_of_docs")
     unprocessed_custom_queries = cfg.opts("workload", "custom_queries")
 
     custom_queries = process_custom_queries(unprocessed_custom_queries)
-    number_of_docs = cfg.opts("generator", "number_of_docs")
 
     logger.info("Creating workload [%s] matching indices [%s]", workload_name, indices)
 

--- a/osbenchmark/workload_generator/workload_generator.py
+++ b/osbenchmark/workload_generator/workload_generator.py
@@ -43,9 +43,11 @@ def process_template(templates_path, template_filename, template_vars, output_pa
         f.write(template.render(template_vars))
 
 
-def extract_mappings_and_corpora(client, output_path, indices_to_extract):
+def extract_mappings_and_corpora(client, output_path, indices_to_extract, total_docs_requested):
     indices = []
     corpora = []
+    docs_were_requested = True if total_docs_requested else False
+
     # first extract index metadata (which is cheap) and defer extracting data to reduce the potential for
     # errors due to invalid index names late in the process.
     for index_name in indices_to_extract:
@@ -55,10 +57,13 @@ def extract_mappings_and_corpora(client, output_path, indices_to_extract):
             logging.getLogger(__name__).exception("Failed to extract index [%s]", index_name)
 
     # That list only contains valid indices (with index patterns already resolved)
-    for i in indices:
-        c = corpus.extract(client, output_path, i["name"])
-        if c:
-            corpora.append(c)
+    for count, idx in enumerate(indices):
+        try:
+            c = corpus.extract(client, output_path, idx["name"], int(total_docs_requested[count]) if docs_were_requested else None)
+            if c:
+                corpora.append(c)
+        except IndexError:
+            raise exceptions.SystemSetupError("--total-docs missing document counts for corresponding indices.")
 
     return indices, corpora
 
@@ -87,6 +92,7 @@ def create_workload(cfg):
     unprocessed_custom_queries = cfg.opts("workload", "custom_queries")
 
     custom_queries = process_custom_queries(unprocessed_custom_queries)
+    total_docs = cfg.opts("generator", "total_docs")
 
     logger.info("Creating workload [%s] matching indices [%s]", workload_name, indices)
 
@@ -99,7 +105,8 @@ def create_workload(cfg):
     output_path = os.path.abspath(os.path.join(io.normalize_path(root_path), workload_name))
     io.ensure_dir(output_path)
 
-    indices, corpora = extract_mappings_and_corpora(client, output_path, indices)
+    indices, corpora = extract_mappings_and_corpora(client, output_path, indices, total_docs)
+
     if len(indices) == 0:
         raise RuntimeError("Failed to extract any indices for workload!")
 

--- a/osbenchmark/workload_generator/workload_generator.py
+++ b/osbenchmark/workload_generator/workload_generator.py
@@ -42,11 +42,23 @@ def process_template(templates_path, template_filename, template_vars, output_pa
     with open(output_path, "w") as f:
         f.write(template.render(template_vars))
 
+def check_if_indices_and_total_docs_match(indices, total_docs, docs_were_requested):
+    '''
+    if --total-docs were provided, the input length needs to be match indices.
+
+    Valid examples: --indices=movies,actors --total-docs=1000,2000
+    Invalid examples: --indices=movies,actors --total-docs=1000 OR
+    --indices=movies --total-docs=1000,2000
+    '''
+    if docs_were_requested and len(indices) != len(total_docs):
+        raise exceptions.SystemSetupError("Both --indices and --total-docs were provided but missing corresponding indices or total docs.")
 
 def extract_mappings_and_corpora(client, output_path, indices_to_extract, total_docs_requested):
     indices = []
     corpora = []
-    docs_were_requested = True if total_docs_requested else False
+    docs_were_requested = total_docs_requested is not None and len(total_docs_requested) > 0
+
+    check_if_indices_and_total_docs_match(indices_to_extract, total_docs_requested, docs_were_requested)
 
     # first extract index metadata (which is cheap) and defer extracting data to reduce the potential for
     # errors due to invalid index names late in the process.
@@ -58,12 +70,9 @@ def extract_mappings_and_corpora(client, output_path, indices_to_extract, total_
 
     # That list only contains valid indices (with index patterns already resolved)
     for count, idx in enumerate(indices):
-        try:
-            c = corpus.extract(client, output_path, idx["name"], int(total_docs_requested[count]) if docs_were_requested else None)
-            if c:
-                corpora.append(c)
-        except IndexError:
-            raise exceptions.SystemSetupError("--total-docs missing document counts for corresponding indices.")
+        c = corpus.extract(client, output_path, idx["name"], int(total_docs_requested[count]) if docs_were_requested else None)
+        if c:
+            corpora.append(c)
 
     return indices, corpora
 

--- a/osbenchmark/workload_generator/workload_generator.py
+++ b/osbenchmark/workload_generator/workload_generator.py
@@ -42,23 +42,23 @@ def process_template(templates_path, template_filename, template_vars, output_pa
     with open(output_path, "w") as f:
         f.write(template.render(template_vars))
 
-def check_if_indices_and_total_docs_match(indices, total_docs, docs_were_requested):
+def check_if_indices_and_number_of_docs_match(indices, number_of_docs, docs_were_requested):
     '''
-    if --total-docs were provided, the input length needs to be match indices.
+    if --number-of-docs were provided, the input length needs to be match indices.
 
-    Valid examples: --indices=movies,actors --total-docs=1000,2000
-    Invalid examples: --indices=movies,actors --total-docs=1000 OR
-    --indices=movies --total-docs=1000,2000
+    Valid examples: --indices=movies,actors --number-of-docs=1000,2000
+    Invalid examples: --indices=movies,actors --number-of-docs=1000 OR
+    --indices=movies --number-of-docs=1000,2000
     '''
-    if docs_were_requested and len(indices) != len(total_docs):
-        raise exceptions.SystemSetupError("Both --indices and --total-docs were provided but missing corresponding indices or total docs.")
+    if docs_were_requested and len(indices) != len(number_of_docs):
+        raise exceptions.SystemSetupError("Both --indices and --number-of-docs were provided but missing corresponding indices or total docs.")
 
-def extract_mappings_and_corpora(client, output_path, indices_to_extract, total_docs_requested):
+def extract_mappings_and_corpora(client, output_path, indices_to_extract, number_of_docs_requested):
     indices = []
     corpora = []
-    docs_were_requested = total_docs_requested is not None and len(total_docs_requested) > 0
+    docs_were_requested = number_of_docs_requested is not None and len(number_of_docs_requested) > 0
 
-    check_if_indices_and_total_docs_match(indices_to_extract, total_docs_requested, docs_were_requested)
+    check_if_indices_and_number_of_docs_match(indices_to_extract, number_of_docs_requested, docs_were_requested)
 
     # first extract index metadata (which is cheap) and defer extracting data to reduce the potential for
     # errors due to invalid index names late in the process.
@@ -70,7 +70,7 @@ def extract_mappings_and_corpora(client, output_path, indices_to_extract, total_
 
     # That list only contains valid indices (with index patterns already resolved)
     for count, idx in enumerate(indices):
-        c = corpus.extract(client, output_path, idx["name"], int(total_docs_requested[count]) if docs_were_requested else None)
+        c = corpus.extract(client, output_path, idx["name"], int(number_of_docs_requested[count]) if docs_were_requested else None)
         if c:
             corpora.append(c)
 
@@ -101,7 +101,7 @@ def create_workload(cfg):
     unprocessed_custom_queries = cfg.opts("workload", "custom_queries")
 
     custom_queries = process_custom_queries(unprocessed_custom_queries)
-    total_docs = cfg.opts("generator", "total_docs")
+    number_of_docs = cfg.opts("generator", "number_of_docs")
 
     logger.info("Creating workload [%s] matching indices [%s]", workload_name, indices)
 
@@ -114,7 +114,7 @@ def create_workload(cfg):
     output_path = os.path.abspath(os.path.join(io.normalize_path(root_path), workload_name))
     io.ensure_dir(output_path)
 
-    indices, corpora = extract_mappings_and_corpora(client, output_path, indices, total_docs)
+    indices, corpora = extract_mappings_and_corpora(client, output_path, indices, number_of_docs)
 
     if len(indices) == 0:
         raise RuntimeError("Failed to extract any indices for workload!")

--- a/osbenchmark/workload_generator/workload_generator.py
+++ b/osbenchmark/workload_generator/workload_generator.py
@@ -48,8 +48,8 @@ def check_if_indices_and_number_of_docs_match(indices, number_of_docs, docs_were
     the number of pairs need to be less than or equal to number of indices
     '''
     if docs_were_requested and len(indices) < len(number_of_docs):
-        raise exceptions.SystemSetupError(f"Number of key:value pairs exceeds number of indices mentioned in --indices. "
-                                          f"Ensure it is less than or equal to.")
+        raise exceptions.SystemSetupError("Number of key:value pairs exceeds number of indices mentioned in --indices. " +
+                                          "Ensure it is less than or equal to.")
 
 def extract_mappings_and_corpora(client, output_path, indices_to_extract, number_of_docs_requested):
     indices = []

